teuthology/suite/__init__.py

```diff
@@ -200,7 +208,106 @@ def get_rerun_conf_overrides(conf):
+            return False
+    else:
+        original_descriptions = [job['description'] for job in run['jobs'] if job['status'] in conf.rerun_statuses and job['description']]
+        conf.filter_in.extend(original_descriptions)
+        log.info("Using backward-compatible job filtering for rerun with %d job descriptions", len(original_descriptions))
+
+
+def _get_default_known_patterns_file():
+    """Returns the path to the default known patterns file bundled with teuthology.
+    
+    The file format is documented in docs/siteconfig.rst.
+    """
+    return os.path.join(os.path.dirname(__file__), 'patterns', 'known-failures.json')
+
+
+def _load_known_patterns(known_patterns_file):
+    """Load known failure patterns from a JSON or YAML file.
+    
+    Args:
+        known_patterns_file: Path to the patterns file. If None, uses the default
+                           bundled file.
+    
+    Returns:
+        List of regex patterns (strings) to match against failure reasons.
+    
+    The file format should be:
+        JSON: {"patterns": ["pattern1", "pattern2", ...]}
+        YAML: patterns: ["pattern1", "pattern2", ...]
+    """
+    if known_patterns_file is None:
+        known_patterns_file = _get_default_known_patterns_file()
+    
+    try:
+        with open(known_patterns_file, 'r') as f:
+            # try JSON first
+            try:
+                patterns_data = json.load(f)
+            except json.JSONDecodeError:
+                # if JSON fails, try YAML
+                f.seek(0)
+                patterns_data = yaml.safe_load(f)
+        
+        known_patterns = patterns_data.get('patterns', [])
+        if not isinstance(known_patterns, list):
+            log.warning(f"Patterns in {known_patterns_file} should be a list, treating as empty")
+            return []
+        return known_patterns
+    except FileNotFoundError:
+        log.warning(f"Could not load known patterns from {known_patterns_file}, treating all as unknown")
+        return []
+    except (yaml.YAMLError, json.JSONDecodeError) as e:
+        log.warning(f"Could not parse known patterns from {known_patterns_file}: {e}, treating all as unknown")
+        return []
+
+
+def filter_out_known_failures(jobs, known_patterns_file=None):
+    """Filter out jobs with known failure patterns.
+    
+    Args:
+        jobs: Iterable of job dictionaries with 'failure_reason' and 'description' keys.
+              Can be a list, generator, or any iterable.
+        known_patterns_file: Optional path to known patterns file. If None, uses
+                           the default bundled file.
+    
+    Returns:
+        List of job dictionaries for jobs with unknown failures (i.e., failures
+        that don't match any known pattern). Only includes jobs that were passed
+        in the input and have descriptions.
+    
+    The known patterns file format is documented in docs/siteconfig.rst.
+    """
```

Comment:
Same as for the function above, would be great to follow teuthology practice of docstring formatting.